### PR TITLE
[mlir][memref] Fix an invalid dim loop motion crash

### LINKED ISSUE #73382: [mlir] --loop-invariant-code-motion crashed with assertion failure.
git version: 024718313b526a4530d7cd58886fcc157247e0ff

system: Ubuntu 20.04.6 LTS (Focal Fossa)

reproduced with: ```mlir-opt --loop-invariant-code-motion a.mlir```

a.mlir:
```
func.func @func2() {
  %c25 = arith.constant 25 : index
  %c22 = arith.constant 22 : index
  %c16 = arith.constant 16 : index
  %c11 = arith.constant 11 : index
  %c0 = arith.constant 0 : index
  %c1 = arith.constant 1 : index
  %c2 = arith.constant 2 : index
  %alloc_13 = memref.alloc(%c16, %c22, %c11) {alignment = 64 : i64} : memref<?x?x?xi1>
  %3 = bufferization.to_tensor %alloc_13 : memref<?x?x?xi1>
  %6 = tosa.arithmetic_right_shift %3, %3 {round = false} : (tensor<?x?x?xi1>, tensor<?x?x?xi1>) -> tensor<?x?x?xi1>
  %13 = scf.for %arg2 = %c0 to %c2 step %c1 iter_args(%arg3 = %3) -> (tensor<?x?x?xi1>) {
    %27 = scf.for %arg4 = %c0 to %c25 step %c1 iter_args(%arg5 = %arg3) -> (tensor<?x?x?xi1>) {
      %dim = memref.dim %alloc_13, %c16 : memref<?x?x?xi1>  // this line cause the crash
      scf.yield %6 : tensor<?x?x?xi1>
    }
    scf.yield %27 : tensor<?x?x?xi1>
  }
  spirv.Return
}
```

trace:
```
mlir-opt: /data/bin/llvm-project/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:934: mlir::Speculation::Speculatability mlir::memref::DimOp::getSpeculatability(): Assertion `constantIndex < rankedSourceType.getRank()' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /data/bin/llvm-project/build/bin/mlir-opt --loop-invariant-code-motion a.mlir
 #0 0x00005569beb4d06f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/data/bin/llvm-project/build/bin/mlir-opt+0x160306f)
 #1 0x00005569beb4a0c4 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f7d2257f420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x00007f7d2206a00b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #4 0x00007f7d22049859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #5 0x00007f7d22049729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #6 0x00007f7d2205afd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #7 0x00005569bfd6327f mlir::memref::DimOp::getSpeculatability() (/data/bin/llvm-project/build/bin/mlir-opt+0x281927f)
 #8 0x00005569bfd20704 mlir::detail::ConditionallySpeculatableInterfaceTraits::Model<mlir::memref::DimOp>::getSpeculatability(mlir::detail::ConditionallySpeculatableInterfaceTraits::Concept const*, mlir::Operation*) (/data/bin/llvm-project/build/bin/mlir-opt+0x27d6704)
 #9 0x00005569c452f921 mlir::isSpeculatable(mlir::Operation*) (/data/bin/llvm-project/build/bin/mlir-opt+0x6fe5921)
#10 0x00005569c180a0f5 mlir::moveLoopInvariantCode(llvm::ArrayRef<mlir::Region*>, llvm::function_ref<bool (mlir::Value, mlir::Region*)>, llvm::function_ref<bool (mlir::Operation*, mlir::Region*)>, llvm::function_ref<void (mlir::Operation*, mlir::Region*)>) (/data/bin/llvm-project/build/bin/mlir-opt+0x42c00f5)
#11 0x00005569c180a80e mlir::moveLoopInvariantCode(mlir::LoopLikeOpInterface) (/data/bin/llvm-project/build/bin/mlir-opt+0x42c080e)
#12 0x00005569c177ef5c void mlir::detail::walk<mlir::ForwardIterator>(mlir::Operation*, llvm::function_ref<void (mlir::Operation*)>, mlir::WalkOrder) (.constprop.2) LoopInvariantCodeMotion.cpp:0:0
#13 0x00005569c177ef5c void mlir::detail::walk<mlir::ForwardIterator>(mlir::Operation*, llvm::function_ref<void (mlir::Operation*)>, mlir::WalkOrder) (.constprop.2) LoopInvariantCodeMotion.cpp:0:0
#14 0x00005569c177fb95 (anonymous namespace)::LoopInvariantCodeMotion::runOnOperation() LoopInvariantCodeMotion.cpp:0:0
#15 0x00005569c174211e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f811e)
#16 0x00005569c17425fa mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f85fa)
#17 0x00005569c1742e24 mlir::PassManager::run(mlir::Operation*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41f8e24)
#18 0x00005569c1733b93 performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#19 0x00005569c1735064 processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPool*) MlirOptMain.cpp:0:0
#20 0x00005569c1735214 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#21 0x00005569c1834504 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/data/bin/llvm-project/build/bin/mlir-opt+0x42ea504)
#22 0x00005569c172da77 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41e3a77)
#23 0x00005569c1735361 mlir::MlirOptMain(int, char**, llvm::StringRef, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41eb361)
#24 0x00005569c17357c6 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41eb7c6)
#25 0x00005569bea845db main (/data/bin/llvm-project/build/bin/mlir-opt+0x153a5db)
#26 0x00007f7d2204b083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#27 0x00005569beb1d3ce _start (/data/bin/llvm-project/build/bin/mlir-opt+0x15d33ce)
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by rikhuijzer**
Fixes https://github.com/llvm/llvm-project/issues/73382.

This PR suggests to replace two assertions that were introduced in https://github.com/llvm/llvm-project/commit/adabce41185910227ca276a1cfd22e76443dd238 (https://reviews.llvm.org/D135748). According to the enum definition of `NotSpeculatable`, an op that invokes undefined behavior is `NotSpeculatable`. 

https://github.com/llvm/llvm-project/blob/0c06e8745f131d867c566f4d35a7a04e24b4a075/mlir/include/mlir/Interfaces/SideEffectInterfaces.h#L248-L258

and both `tensor.dim` and `memref.dim` state that "If the dimension index is out of bounds, the behavior is undefined."

So therefore it seems to me that `DimOp::getSpeculatability()` should return `NotSpeculatable` if the dimension index is out of bounds.

The added test is just a simplified version of https://github.com/llvm/llvm-project/issues/73382.

### CODE REVIEW

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp [hunk 1]
```diff
@@ -893,8 +893,9 @@ Speculation::Speculatability DimOp::getSpeculatability() {
   if (!rankedSourceType)
     return Speculation::NotSpeculatable;
 
-  // The verifier rejects operations that violate this assertion.
-  assert(constantIndex < rankedSourceType.getRank());
+  if (rankedSourceType.getRank() < constantIndex)
```
Comment (sanjoy) [inline mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp:896]: Is the comment factually incorrect: `The verifier rejects operations that violate this assertion.`?

Also, I think the condition should be `rankedSourceType.getRank() <= constantIndex`.  Same for the other check.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp (hunk 1 above)
Comment (rikhuijzer) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp]: > Is the comment factually incorrect: `The verifier rejects operations that violate this assertion.`?

The comment is correct. I think what happened is that the idea about this has changed over time, the assertion was added in Oct 2022, but an argument to not crash was made Jan 2023, see https://github.com/llvm/llvm-project/issues/60295. From my interpretation of https://github.com/llvm/llvm-project/issues/60295, the comment is correct but crashing is not. 

> Also, I think the condition should be `rankedSourceType.getRank() <= constantIndex`. Same for the other check.

Well spotted! I've added tests in [53e2bcd](https://github.com/llvm/llvm-project/pull/74204/commits/53e2bcd0cc42b518d45f7d8c71373b3da633ec4a) to ensure that `-loop-invariant-code-motion` optimizes the code when it should.

Comment (joker-eph) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp]: Can you clarify why the verifier shouldn't check this invariant Sanjoy? 
Reading this quickly, it seems like a purely local invariant?

Comment (joker-eph) [reply mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp]: Seems right to me.

**Review by joker-eph — APPROVED** [review]

**Review by sanjoy — APPROVED** [review]

**Comment by rikhuijzer** [comment]
Thanks both for the review!

**Pull request merged**